src/lib/buySellHelper/buySellHelper.ts

```diff
@@ -0,0 +1,84 @@
+import { TradesDataType, TRADE_STATUS } from "@jaws/db/tradesMeta";
+import { isToday } from "@jaws/lib/helpers";
+import { BuySellConstants, DEFAULT_BUY_SELL_CONFIG } from "./buySellConstants";
+
+export const getBuySellHelpers = (config?: Partial<BuySellConstants>) => {
+  const _config = { ...DEFAULT_BUY_SELL_CONFIG, ...config };
+
+  return {
+    get config() {
+      return { ..._config } as const;
+    },
+
+    /**
+     * Returns the maximum amount of money (total value) the position is
+     * allowed to drop before triggering Stop loss (1)
+     */
+    getStopLossMaxAmount: (totalAssets: number): number =>
+      totalAssets * _config.STOP_LOSS_1_PORTFOLIO_PERCENTAGE,
+
+    determineTradeStatus: (opts: {
+      trade: TradesDataType;
+      lastTradePrice: number;
+      movingAvg: number;
+      stopLossMaxAmount: number;
+    }): TRADE_STATUS | undefined => {
+      const stopLossType = determineStopLossType(opts);
+
+      if (stopLossType) {
+        return stopLossType;
+      } else if (isTakePartialProfit(opts)) {
+        return TRADE_STATUS.TAKE_PARTIAL_PROFIT;
+      }
+
+      return;
+    },
+  };
+
+  function determineStopLossType({
+    trade,
+    stopLossMaxAmount,
+    lastTradePrice,
+    movingAvg,
+  }: {
+    trade: TradesDataType;
+    stopLossMaxAmount: number;
+    lastTradePrice: number;
+    movingAvg: number;
+  }): TRADE_STATUS | undefined {
+    // Stop loss case (1)
+    if ((trade.price - lastTradePrice) * trade.quantity >= stopLossMaxAmount)
+      return TRADE_STATUS.STOP_LOSS_1;
+
+    if (!isToday(trade.created)) {
+      // Stop loss case (2)
+      if (lastTradePrice <= trade.price) return TRADE_STATUS.STOP_LOSS_2;
+
+      // Stop loss case (3) Take profit
+      if (movingAvg && lastTradePrice <= movingAvg)
+        return TRADE_STATUS.STOP_LOSS_3;
+    }
+
+    return;
+  }
+
+  /** After 10% increase in value, we take profit */
+  function isTakePartialProfit({
+    trade,
+    lastTradePrice,
+  }: {
+    trade: TradesDataType;
+    lastTradePrice: number;
+  }) {
+    if (trade.status === TRADE_STATUS.TAKE_PARTIAL_PROFIT) {
```

Comment:
I guess this name, TAKE_PARTIAL_PROFIT, was determined a long time ago, but what does it mean really? Here it seems like it means that a partial profit has been taken, while the name, in the imperative form, suggests that an action should be taken 🤔 Maybe it should be PARTIAL_PROFIT_TAKEN, sounds more like status to me. And then we should not set the trade in this state until the profit is taken of course. I guess now this might status will trigger a trade? We could instead check for the condition in the CRON, take action, and set status if succeded?

This might not make sense to you, as I am making a lot of assumption on how I think things work, in that case just ignore me :)

Reply:
You're right, it is applied after the partial profit has been taken here: https://github.com/hejare/jaws/blob/193-trades-break-out-magic-numbers-to-a-separate-config/src/lib/tradesHandler.ts#L243

I like your suggestion, will change it. As for whether the order is completed in Alpaca or not, at this point we just assume everything is executed as we want to, and instantly. :^)